### PR TITLE
fix: use iterables instead of generators

### PR DESCRIPTION
**Pull request by jakebailey**
This applies https://github.com/mswjs/msw/pull/2108#discussion_r1685551423 to the `fix/generator-resolver` branch.

Best viewed without whitespace: https://github.com/mswjs/msw/pull/2213/files?w=1

### CODE REVIEW

File: src/core/handlers/RequestHandler.ts [hunk 1]
```diff
@@ -314,37 +318,38 @@ export abstract class RequestHandler<
...
-        this.isUsed = false
-
-        if (!this.resolverGenerator) {
-          this.resolverGenerator = result
+      if (!this.resolverGenerator) {
+        const result = await resolver(info)
+        if (!isIterable(result)) {
+          return result
         }
+        this.resolverGenerator =
```
Comment (jakebailey) [inline src/core/handlers/RequestHandler.ts]: Technically, this isn't a generator anymore, so the variable name is a lie.

File: src/core/handlers/RequestHandler.ts [hunk 2]
```diff
@@ -314,37 +318,38 @@ export abstract class RequestHandler<
...
+          Symbol.iterator in result
```
Comment (jakebailey) [inline src/core/handlers/RequestHandler.ts:327]: `Reflect.has` (used in `isIterable`) is the same operation as `in`, except that TS can acutally narrow using `in`.

File: src/core/handlers/RequestHandler.ts (hunk 1 above)
Comment (kettanaito) [reply src/core/handlers/RequestHandler.ts]: Is something like `resolverIterator` a more fitting name? These are internal so we can rename them to our hearts' content.

File: src/core/utils/internal/isIterable.ts [hunk 3]
```diff
@@ -1,11 +1,20 @@
+// Redefine these types; the three-parameter forms are only available in TypeScript 5.6 or later.
```
Comment (kettanaito) [inline src/core/utils/internal/isIterable.ts]: Thank you for putting this in! Exactly what I was wondering about.

**Comment by kettanaito** [comment]
😱 

Screenshot 2024-07-21 at 19 35 24

File: src/core/utils/internal/isIterable.ts (hunk 3 above)
Comment (jakebailey) [reply src/core/utils/internal/isIterable.ts]: I can also make these JSDoc comments so they're carried down; whatever you'd prefer.

Comment (kettanaito) [reply src/core/utils/internal/isIterable.ts]: I'd mark them as `@todo` to remove these types once TS 5.6 is the minimal supported version. Not going to happen any time soon but still.

Comment (jakebailey) [reply src/core/utils/internal/isIterable.ts]: Updated.

**Review by kettanaito — APPROVED** [review]
This looks incredible! 🎉 Thank you so much, @jakebailey. You've made the impossible possible. Local tests are passing, type-compatibility on CI is green.

**Pull request merged**